apps/passport-server/migrations/61_rate_limit_variable_periods.sql

```diff
@@ -0,0 +1,63 @@
+-- Moving rate limit configuration to TypeScript code, so this is no longer
+-- needed.
+ALTER TABLE rate_limit_buckets DROP CONSTRAINT rate_limit_buckets_action_type_fkey;
+DROP TABLE rate_limit_types;
+
+-- Account reset timestamp column no longer required
+ALTER TABLE users DROP COLUMN account_reset_timestamps;
+
+-- Re-implementation of function first created in 58_rate_limit.sql
+-- This is simplified by allowing more values to be passed in from the calling
+-- side, instead of relying on lookups from the rate_limit_types table.
+--
+-- This means that the caller (RateLimitService) can manage its own
+-- configuration of which types of action exist, how many actions are allowed,
+-- and what the time period is.
+-- See the `limits` private member variable in `RateLimitService`.
+DROP FUNCTION IF EXISTS take_token(VARCHAR, VARCHAR, TIMESTAMPTZ);
+CREATE OR REPLACE FUNCTION take_token(
+  action_type VARCHAR, -- The type of action, such as "ACCOUNT_RESET"
+  action_id VARCHAR, -- A unique identifier, such as the account ID being reset
+  periodic_limit INTEGER, -- How many of these actions are permitted per period
+  time_period_seconds INTEGER, -- The time period, in seconds
+  time_now TIMESTAMPTZ) -- The current time
+  RETURNS boolean AS $$
+DECLARE
+    tokens INTEGER;
+    extra_tokens INTEGER;
+    new_tokens INTEGER;
+    last_refill TIMESTAMPTZ;
+    this_refill TIMESTAMPTZ;
+BEGIN
+    -- Lock the buckets until end of transaction
+    LOCK TABLE rate_limit_buckets IN EXCLUSIVE MODE;
```

Comment:
Locking the whole table, as opposed to a single row, seems like it could cause contention problems if we use rate limiting for something high-rate, like request/transaction limits.  Would a SELECT FOR UPDATE below be sufficient?